activesupport/lib/active_support/message_verifier.rb

```diff
@@ -64,54 +66,162 @@ def valid_message?(signed_message)
+    #
+    # JWS Signature - Signature over the JWS Header and the JWS Payload(Claims set).
+    # The signing input is given by ASCII(BASE64URL(UTF8(JWS Protected Header)) || '.' ||
+    # BASE64URL(JWS Payload)).
+    #
+    # ==== Options
+    #
+    # * <tt>:expires_at</tt> - Expiry time of the signed message.
+    # * <tt>:expires_in</tt> - Time from now after which the message
+    #   will expire (e.g. 1.month).
+    # * <tt>:for</tt> - Purpose/Scope of the signed message (defaults to 'universal').
+    #
+    # ==== Examples
+    #
+    # Messages can be signed to ensure that the data hasn't been tampered with.
     #
     #   verifier = ActiveSupport::MessageVerifier.new 's3Krit'
-    #   verifier.generate 'a private message' # => "BAhJIhRwcml2YXRlLW1lc3NhZ2UGOgZFVA==--e2d724331ebdee96a10fb99b089508d1c72bd772"
-    def generate(value)
-      data = encode(@serializer.dump(value))
-      "#{data}--#{generate_digest(data)}"
+    #   verifier.generate('a private message')
+    #   # => "BAh7B0kiCHR5cAY6BkVUSSIISldUBjsAVEkiCGFsZwY7AFRJIglTSEExBjsAVA==.BAh7CDoIcGxkSSIWYSBwcml2YXRlIG1lc3NhZ2UGOgZFVDoIZm9ySSIOdW5pdmVyc2FsBjsGVDoIZXhwSSIdMjAxNS0wOC0wOVQxODowNjowMy44NzhaBjsGVA==.ZmRlYjhlMTkxNGI0NGQ4OWRkNDdjZWM5YTY1NmQ3MTM5OWY5NzRmMQ=="
+    #
+    # Messages can be given a purpose/scope to bump the security up some more.
+    # In this way evildoers can't reuse a sign-up form's cookie on the login
+    # page. For example:
+    #
+    #   signup_message = verifier.generate('a private message', for: 'signup_form')
+    #   # => "BAh7B0kiCHR5cAY6BkVUSSIISldUBjsAVEkiCGFsZwY7AFRJIglTSEExBjsAVA==.BAh7CDoIcGxkSSIWYSBwcml2YXRlIG1lc3NhZ2UGOgZFVDoIZm9ySSIQc2lnbnVwX2Zvcm0GOwZUOghleHBJIh0yMDE1LTA4LTA5VDE4OjA1OjEwLjM2N1oGOwZU.YTMxMzIzMDc3YTI3ZDkwM2FmNWQ1MDgwNzAxNTgwMjNjNzI2NDE0Zg=="
+    #
+    #   verifier.verify(signup_message, for: 'signup_form')
+    #   # => 'a private message'
+    #
+    # A message can be set that expires some time in future with +expires_in+.
+    # This is useful for cookies/messages which should not be indefinitely
+    # accessible. For example:
+    #
+    #   expiring_message = verifier.generate('a private message', for: 'remember_me', expires_in: 1.hour)
+    #   # => "BAh7B0kiCHR5cAY6BkVUSSIISldUBjsAVEkiCGFsZwY7AFRJIglTSEExBjsAVA==.BAh7CDoIcGxkSSIWYSBwcml2YXRlIG1lc3NhZ2UGOgZFVDoIZm9ySSIQcmVtZW1iZXJfbWUGOwZUOghleHBJIh0yMDE1LTA3LTA5VDE5OjAzOjEwLjQ3MVoGOwZU.MmQxNzkxMmY2MGU5ZmQ2YTlhNmRhYmYyNmEzNDUzNjE0ZDdjNzg2Yw=="
+    #
+    #   # Within 1 hour...
+    #   verifier.verify(expiring_message, for: 'remember_me')
+    #   # => "a private message"
+    #
+    #   # After 1 hour...
+    #   verifier.verify(expiring_message, for: 'remember_me')
+    #   # raises ActiveSupport::Claims::ExpiredClaims
+    #
+    # A message can also be set to expire at an explicit time with +expires_at+:
+    #
+    #   expiring_message = verifier.generate('a private message', for: 'something', expires_at: 2.hours.from_now)
```

Comment:
Using `from_now` makes this feel like the `expires_in` example. Let's use `Date.tomorrow.midnight`. Then...